### ISSUE TITLE
Conditionally evaluate expression with aux or provide a dummy output

### ISSUE THREAD

**Issue by hamzamerzic** [google/jax#2660]
I have a function that I want to evaluate given a condition. The function returns a loss and extra (or aux in JAX world :)) or dummy values if the condition is not satisfied. 
Basically, I’m trying to do something like:
```python
counter = jnp.array(5)
result = jax.lax.cond(
    pred=counter >= 10,
    true_operand=(1., dict(a=3, b=4)),  # Extra can be unknown apriori.
    true_fun=lambda x: x,
    false_operand=(0., dict()),  # Dummy.
    false_fun=lambda x: x)
```
but this fails due to a check that the extra in `false_operand` doesn’t contain keys `a` and `b`. Is there any way I can disable this nested check? 

I could of course specify the same structure, but I would like to be able to have different losses as `true_fun`'s, or aggregate of losses and combination of the corresponding extra values.

**Comment by hawkinsp** [comment]
The easiest solution is to try using regular Python control flow i.e., (`if ...`) instead of using `lax.cond`. This only works if you don't need to trace through the conditional evaluation (e.g., to stage it into a `jit` or `pmap`). In some applications, this works fine.

Also, if `true_fun` is cheap, you might consider just evaluating `true_fun` unconditionally and ignoring the value it produces in some cases?

If not, I think for now your best solution is to mimic the tree/array structure on both branches. This is somewhat fundamental, in that the output of `cond` must have a single, well-defined type irrespective of which branch is taken.

If we had a concept of sum types in JAX, it might help here, but we don't at the moment.

What do you think?

**Comment by hamzamerzic** [comment]
Thanks for the detailed reply! I went with the second option of always evaluating `true_fun`, but for the future it would defintely save compute to be able to just return dummy values in the `false` branch.

**Comment by mattjj** [comment]
Closing as old, and mostly answered!
